### PR TITLE
Fix: cpu 100% in session.handleLoop

### PR DESCRIPTION
**Pull request by pantianying**

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -856,5 +857,5 @@ func (s *session) gc() {
 func (s *session) Close() {
 	s.stop()
 	log.Info("%s closed now. its current gr num is %d",
```
Comment (wongoo) [inline session.go]: need update log content

File: session.go [hunk 2]
```diff
@@ -577,7 +578,7 @@ func (s *session) handlePackage() {
 			log.Errorf("[session.handlePackage] panic session %s: err=%s\n%s", s.sessionToken(), r, rBuf)
 		}
 
-		grNum = atomic.AddInt32(&(s.grNum), -1)
+		close(s.handlePackageDone)
 		log.Infof("%s, [session.handlePackage] gr will exit now, left gr num %d", s.sessionToken(), grNum)
```
Comment (wongoo) [inline session.go]: need update log content

File: session.go [hunk 3]
```diff
@@ -496,7 +499,7 @@ func (s *session) handleLoop() {
 			log.Errorf("[session.handleLoop] panic session %s: err=%s\n%s", s.sessionToken(), r, rBuf)
 		}
 
-		grNum = atomic.AddInt32(&(s.grNum), -1)
+		close(s.handleLoopDone)
 		s.listener.OnClose(s)
 		log.Info("%s, [session.handleLoop] goroutine exit now, left gr num %d", s.Stat(), grNum)
```
Comment (fangyincheng) [inline session.go]: delete `grNum`

File: session.go [hunk 4]
```diff
@@ -102,9 +103,11 @@ func newSession(endPoint EndPoint, conn Connection) *session {
 
 		period: period,
 
-		done:  make(chan struct{}),
-		wait:  pendingDuration,
-		attrs: NewValuesContext(nil),
+		done:              make(chan struct{}),
+		wait:              pendingDuration,
+		attrs:             NewValuesContext(nil),
+		handleLoopDone:    make(chan struct{}),
```
Comment (fangyincheng) [inline session.go]: the `handleLoopDone`   is not used, can it be deleted?

**Comment by wongoo** [comment]
Is this PR really fix cpu 100% issue?  
The info I got is that the value of `s.grNum` not changing to 1 causes `handleLoop()` can not break and go into a dead loop. is it correct?   
And the community performance test result says that `handlePackage` may blocked on read and won't exit. Is that correct?

**Comment by AlexStocks** [comment]
> Is this PR really fix cpu 100% issue?
> The info I got is that the value of `s.grNum` not changing to 1 causes `handleLoop()` can not break and go into a dead loop. is it correct?
> And the community performance test result says that `handlePackage` may blocked on read and won't exit. Is that correct?

Yes, it is for cpu 100% bug fix. The session.handleLoop will run very quickly when the ss.Done is closed but the done select branch can not block when the handlePackage has not exited. And panty add the session.rDone to let the session.handleLoop for-select-done branch block on '<-session.rDone'.

the "`handlePackage` may blocked on read and won't exit." case is that the handlePackage will try to reconnect the server when it invokes the session.stop(). At last it will exit when it connects the server successfully or this session is closed by upper invoker. However, this case does not matter because it should not block the handleLoop to wait.

**Comment by AlexStocks** [comment]
LGTM

**Comment by wongoo** [comment]
LGTM

**Pull request merged**